### PR TITLE
MGDAPI-4257 Adding custom domain

### PR DESCRIPTION
**Pull request by Boomatang**
# Issue link
https://issues.redhat.com/browse/MGDAPI-4257

# What

This work allows the configuration of a custom domain name for 3scale. 

## Still to do
* dev guide on how to configure custom domain in the cluster
* forwards alerts to the BU/customer
* document the e2e test required

# Verification steps

For testing the installation may be done by olm. Metrics from the operator are required in the alerts.
This [INDEX](quay.io/jfitzpat/managed-api-service-index:1.24.0-4257) can be used.

Use the [Guide](https://github.com/integr8ly/integreatly-operator/blob/599642dd50b024e56bfc27fc38e8978a3e20f09f/guides/custom_domain.md) which is part of this PR to help with configuring the custom domain feature. The guide will configure the custom domain in the correct manner. For the verification the installation will need to be broken. Adjust the guide for the installation step that is currently in progress. 

## Bad addon parameter value
**Step up**
When setting the `custom-domain_domain` in the addon secret. 
Use a invalid domain name. example `bad domain`.

**Expected**
* RHMI CR status should give error on the bad domain. 
`oc get RHMI rhoam -n redhat-rhoam-operator -o jsonpath='{.status.customDomain.error}{"\n"}'`
Expected error: "not valid domain "bad domain""
* Once the monitoring stack is installed, the `CustomDomainCRErrorState` should be pending or firing
* Install should stop before installing 3scale
 
## Missing custom domain CR
**Step up**
When setting the `custom-domain_domain` in the addon secret. 
Use a valid domain name but do not create the customDomain CR. 

**Expected**
* RHMI CR status should give error on the bad domain. 
`oc get RHMI rhoam -n redhat-rhoam-operator -o jsonpath='{.status.customDomain.error}{"\n"}'`
Expected error: "finding CustomDomain CR failed: no custom domain CR found for: "apps.rhmi.me""
* Once the monitoring stack is installed, the `CustomDomainCRErrorState` should be pending or firing
* Install should stop before installing 3scale

## Not ready state custom domain CR
**Step up**
Create a CustomDomain CR that has the correct domain URL but is stuck in a non Ready state. 
Easiest way to do this is by adding the wrong to certificate name in the CR spec.

**Expected**
* RHMI CR status should give error on the bad domain. 
`oc get RHMI rhoam -n redhat-rhoam-operator -o jsonpath='{.status.customDomain.error}{"\n"}'`
Expected error: "finding CustomDomain CR failed: custom domain CR in failing state for: "apps.rhmi.me""
* Once the monitoring stack is installed, the `CustomDomainCRErrorState` should be pending or firing
* Install should stop before installing 3scale

## Correct installation
**Step Up**
Allow the install to come to a complete state.
This would mean fixing any issues cause while doing the earlier testing.

Set up the testing IDP and log into 3scale as a test user.

**Expected** 
* 3Scale login should happen and be redirected to custom domain URL
* routes in 3Scale namespace all have custom domain names.

### CODE REVIEW

File: guides/custom_domain.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+-n redhat-rhoam-operator --type=json \
+-p='[{"op": "replace", "path": /data/custom-domain_domain, "value": "{}"}]'
+```
+As this setup is using self-signed certificates the operator will need to be configured to allow this.
+One way of doing this is to create the rhmi CR on cluster with the make target of `make deploy/integreatly-rhmi-cr.yml`.
+If the rhmi CR has already been created it can be patched after the installation has started. 
+Patch the CR as follows.
+
+```shell    
+oc patch rhmi rhoam -n redhat-rhoam-operator --type=json -p='[{"op": "replace", "path": /spec/selfSignedCerts, "value": "true"}]'
```
Comment (cecobask) [inline guides/custom_domain.md:88]:
```suggestion
oc patch rhmi rhoam -n redhat-rhoam-operator --type=json -p='[{"op": "replace", "path": /spec/selfSignedCerts, "value": true}]'
```
```suggestion
oc patch rhmi rhoam -n redhat-rhoam-operator --type=json -p='[{"op": "replace", "path": /spec/selfSignedCerts, "value": "true"}]'
```

File: guides/custom_domain.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+This guide does not cover the steps required to configure the CNAME record in the DNS configuration.
+
+# Configuring the operator installation.
+Currently, the operator only allows the configuring of the custom domain during installation. 
+Once the routes have being created within 3scale they can not be changed again.
+
+Run the normal cluster prepare step, `make cluster/prepare/local`.
+Patch the secret with the domain name.
+```shell
+echo -n "apps.rhmi.com" | base64 | xargs -I {} \
```
Comment (cecobask) [inline guides/custom_domain.md:77]:
```suggestion
echo -n "apps.rhmi.me" | base64 | xargs -I {} \
```
```suggestion
echo -n "apps.rhmi.com" | base64 | xargs -I {} \
```

Comment (cecobask) [inline guides/custom_domain.md:77]:
```suggestion
echo -n "apps.rhmi.me" | base64 | xargs -I {} \
```

File: guides/custom_domain.md [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+    name: rhmi-me-tls
+    namespace: certs
+EOF
+```
+
+Once the Custom Domain CR is in a ready state there will be an endpoint value that is required by the CNAME record.
+
+Checking the state of the Custom Domain CR.
+```shell
+oc get CustomDomain customdoamin1 -o jsonpath='{.status.state}{"\n"}'
```
Comment (cecobask) [inline guides/custom_domain.md:58]:
```suggestion
oc get CustomDomain customdomain1 -o jsonpath='{.status.state}{"\n"}'
```

**Review by cecobask — COMMENTED** [review]
Tested the happy path separately on a clean cluster.
3scale console still not accessible, but this time the error from the keycloak pod in rhsso namespace is the following:
```
Caused by: java.net.UnknownHostException: api.rhmi.me: Name or service not known
```
I made sure that I entered the name `apps.rhmi.me` as the custom domain name everywhere, so not sure where `api.rhmi.me` is coming from.

File: pkg/resources/custom-domain/custom-domain.go [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+	customDomains := &customdomainv1alpha1.CustomDomainList{}
+
+	err := serverClient.List(ctx, customDomains)
+	if err != nil {
+		return false, err
+	}
+
+	for _, item := range customDomains.Items {
+		if item.Spec.Domain == domain {
+			if item.Status.State == "Ready" {
```
Comment (adam-cattermole) [inline pkg/resources/custom-domain/custom-domain.go:73]: Minor nit
```suggestion
			if item.Status.State == customdomainv1alpha1.CustomDomainStateReady {
```

**Comment by Boomatang** [comment]
> The only other thing is the error handling TODO in the bootstrapReconciler, should the reconcile fail if an invalid custom domain is set to prevent continuing?

@adam-cattermole The reconcile does fail if there is an invalid custom domain. The failure happens in the 3scale reconciler. Main reason for doing it in the 3scale reconciler over the bootstrap reconciler is to allow use to use own monitoring stack and not need to send these failure alerts to SRE-P. Does this answer your question.

**Comment by adam-cattermole** [comment]
Sorry @Boomatang my question was worded incorrectly.

I was more mentioning your TODO message within the bootstrapReconciler L619. If for example the addon parameter is not retrieved successfully and an error is returned we log the error but continue to set the domain as the default one. I know it's pretty unlikely that the read of the k8s API fails but in this case should we return phase failed and wait until there is no error or is it fine as is where we log the error and continue?

**Comment by Boomatang** [comment]
@adam-cattermole Now I see what you where talking about. That TODO I forgot to remove. It was a comment from when I was starting. I do agree we should return a failed state if there is an unexpected error from ` customDomain.GetDomain`, These errors would likely be due to k8s API failures in which case we would want to retry. I will make an update to address this.

**Comment by adam-cattermole** [comment]
hmm timed out

/test rhoam-e2e

**Comment by adam-cattermole** [comment]
I thought this was just being really flaky but I've had a look at the artifacts and it looks like operator pod is stuck retrieving the apiServerURL:

```
2022-07-21T16:46:23.818Z	DEBUG	controller-runtime.manager.events	Warning	{"object": {"kind":"RHMI","namespace":"redhat-rhoam-operator","name":"rhoam","uid":"861517dc-8a90-4944-be30-f585db05f68f","apiVersion":"integreatly.org/v1alpha1","resourceVersion":"34476"}, "reason": "ProcessingError", "message": "Failed to retrieve API server URL:\nno ServingCerts.NamedCertificates found starting with api"}
```

The bootstrap sits at phasefailed and the e2e never completes. I'm wondering if there's some other potential format for the apiserver CR certs URL.

**Comment by adam-cattermole** [comment]
I've dug a little more into this.

It looks like the apiserver CR has no certs:

```
apiVersion: config.openshift.io/v1
items:
- apiVersion: config.openshift.io/v1
  kind: Image
  metadata:
    annotations:
      include.release.openshift.io/ibm-cloud-managed: "true"
      include.release.openshift.io/self-managed-high-availability: "true"
      include.release.openshift.io/single-node-developer: "true"
      release.openshift.io/create-only: "true"
    creationTimestamp: "2022-07-21T16:08:06Z"
    generation: 1
    managedFields:
    - apiVersion: config.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:metadata:
          f:annotations:
            .: {}
            f:include.release.openshift.io/ibm-cloud-managed: {}
            f:include.release.openshift.io/self-managed-high-availability: {}
            f:include.release.openshift.io/single-node-developer: {}
            f:release.openshift.io/create-only: {}
          f:ownerReferences:
            .: {}
            k:{"uid":"123b7244-30a1-4ea6-ae40-0ccdfd191404"}: {}
        f:spec: {}
      manager: cluster-version-operator
      operation: Update
      time: "2022-07-21T16:08:06Z"
    - apiVersion: config.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          .: {}
          f:internalRegistryHostname: {}
      manager: Go-http-client
      operation: Update
      subresource: status
      time: "2022-07-21T16:15:38Z"
    name: cluster
    ownerReferences:
    - apiVersion: config.openshift.io/v1
      kind: ClusterVersion
      name: version
      uid: 123b7244-30a1-4ea6-ae40-0ccdfd191404
    resourceVersion: "19116"
    uid: 11f38cac-9693-434c-81f8-e2a4d1ff2ca5
  spec: {}
  status:
    internalRegistryHostname: image-registry.openshift-image-registry.svc:5000
kind: ImageList
metadata:
  continue: ""
  resourceVersion: "43870"
```

but the infrastructure CR does have the api url, so perhaps we should use that instead?

```
apiVersion: config.openshift.io/v1
items:
- apiVersion: config.openshift.io/v1
  kind: Infrastructure
  metadata:
    creationTimestamp: "2022-07-21T16:07:28Z"
    generation: 1
    managedFields:
    - apiVersion: config.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:spec:
          .: {}
          f:cloudConfig:
            .: {}
            f:name: {}
          f:platformSpec:
            .: {}
            f:aws: {}
            f:type: {}
      manager: cluster-bootstrap
      operation: Update
      time: "2022-07-21T16:07:28Z"
    - apiVersion: config.openshift.io/v1
      fieldsType: FieldsV1
      fieldsV1:
        f:status:
          .: {}
          f:apiServerInternalURI: {}
          f:apiServerURL: {}
          f:controlPlaneTopology: {}
          f:etcdDiscoveryDomain: {}
          f:infrastructureName: {}
          f:infrastructureTopology: {}
          f:platform: {}
          f:platformStatus:
            .: {}
            f:aws:
              .: {}
              f:region: {}
            f:type: {}
      manager: cluster-bootstrap
      operation: Update
      subresource: status
      time: "2022-07-21T16:07:28Z"
    name: cluster
    resourceVersion: "602"
    uid: 474f38c1-63af-4044-949b-57f3b5a60bc2
  spec:
    cloudConfig:
      name: ""
    platformSpec:
      aws: {}
      type: AWS
  status:
    apiServerInternalURI: https://api-int.ci-ocp-4-10-amd64-aws-us-west-1-5ff4z.hive.aws.ci.openshift.org:6443
    apiServerURL: https://api.ci-ocp-4-10-amd64-aws-us-west-1-5ff4z.hive.aws.ci.openshift.org:6443
    controlPlaneTopology: HighlyAvailable
    etcdDiscoveryDomain: ""
    infrastructureName: ci-ocp-4-10-amd64-aws-vf64w
    infrastructureTopology: HighlyAvailable
    platform: AWS
    platformStatus:
      aws:
        region: us-west-1
      type: AWS
kind: InfrastructureList
metadata:
  continue: ""
  resourceVersion: "43871"

```

**Comment by adam-cattermole** [comment]
Looks like the servingCerts field is optional so I guess it could be a problem https://docs.openshift.com/container-platform/4.10/rest_api/config_apis/apiserver-config-openshift-io-v1.html

**Review by cecobask — APPROVED** [review]
Changes look good to me. Adam's comments were addressed.

/lgtm

**Pull request merged**